### PR TITLE
Sema: implement @splat for arrays

### PR DESCRIPTION
**Pull request by mlugg**
Resolves: #20433

### CODE REVIEW

File: src/Sema.zig [hunk 1]
```diff
@@ -8615,8 +8615,9 @@ fn zirVectorElemType(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileEr
         error.GenericPoison => return .generic_poison_type,
         else => |e| return e,
     };
-    if (!vec_ty.isVector(zcu)) {
-        return sema.fail(block, block.nodeOffset(un_node.src_node), "expected vector type, found '{}'", .{vec_ty.fmt(pt)});
+    switch (vec_ty.zigTypeTag(zcu)) {
+        .array, .vector => {},
+        else => return sema.fail(block, block.nodeOffset(un_node.src_node), "expected vector type, found '{}'", .{vec_ty.fmt(pt)}),
```
Comment (Rexicon226) [inline src/Sema.zig:8620]:
```suggestion
        else => return sema.fail(block, block.nodeOffset(un_node.src_node), "expected array or vector type, found '{}'", .{vec_ty.fmt(pt)}),
```
did you mean?

**Review by Rexicon226 — APPROVED** [review]

File: src/Sema.zig [hunk 2]
```diff
@@ -24804,26 +24805,65 @@ fn zirSplat(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
...
+                .storage = s: {
+                    full: {
+                        if (dest_ty.zigTypeTag(zcu) == .vector) break :full;
+                        const sentinel = maybe_sentinel orelse break :full;
+                        if (sentinel.toIntern() == scalar_val.toIntern()) break :full;
+                        const len = dest_ty.arrayLen(zcu);
+                        if (len == 0) break :full;
+                        // This is a array with non-zero length and a sentinel which does not match the element.
+                        // We have to use the full `elems` representation.
+                        const elems = try sema.arena.alloc(InternPool.Index, len + 1);
```
Comment (Rexicon226) [inline src/Sema.zig:24843]:
```suggestion
                        const elems = try sema.arena.alloc(InternPool.Index, @intCast(len + 1));
```
Maybe `arrayLen` should return `usize`? I've had to intcast the length from it so many times. vectors take a `u32` len fwiw.

File: src/Sema.zig [hunk 3]
```diff
@@ -24804,26 +24805,65 @@ fn zirSplat(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
...
+                        if (dest_ty.zigTypeTag(zcu) == .vector) break :full;
+                        const sentinel = maybe_sentinel orelse break :full;
+                        if (sentinel.toIntern() == scalar_val.toIntern()) break :full;
+                        const len = dest_ty.arrayLen(zcu);
+                        if (len == 0) break :full;
+                        // This is a array with non-zero length and a sentinel which does not match the element.
+                        // We have to use the full `elems` representation.
+                        const elems = try sema.arena.alloc(InternPool.Index, @intCast(len + 1));
+                        @memset(elems[0..len], scalar_val.toIntern());
+                        elems[len] = sentinel.toIntern();
```
Comment (Rexicon226) [inline src/Sema.zig:24845]: these need to be `@intCast`ed, prob just cast the `const len = `

File: src/Sema.zig [hunk 4]
```diff
@@ -24804,26 +24805,65 @@ fn zirSplat(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
...
     }
 
     try sema.requireRuntimeBlock(block, src, scalar_src);
-    return block.addTyOp(.splat, dest_ty, scalar);
+
+    switch (dest_ty.zigTypeTag(zcu)) {
+        .array => {
+            const len = dest_ty.arrayLen(zcu);
+            const elems = try sema.arena.alloc(Air.Inst.Ref, len + @intFromBool(maybe_sentinel != null));
+            @memset(elems[0..len], scalar);
```
Comment (Rexicon226) [inline src/Sema.zig:24860]: these two will also need to be int casted. ditto above.

**Review by Rexicon226 — CHANGES_REQUESTED** [review]

**Comment by andrewrk** [comment]
Release notes please!

**Comment by mlugg** [comment]
## Release Notes

Zig 0.14.0 expands the `@splat` builtin to apply not only to vectors, but to arrays. This is useful when default-initializing an array to a constant value. For instance, in conjunction with \[Decl Literals\}\(link to relevent release notes section\), we can elegantly initialize an array of "color" values:
```zig
const Rgba = struct {
    r: u8,
    b: u8,
    g: u8,
    a: u8,
    pub const black: Rgba = .{ .r = 0, .g = 0, .b = 0, .a = 255 };
};
var pixels: [width][height]Rgba = @splat(@splat(.black));
```
The operand may be comptime-known or runtime-known. In addition, this builtin can also be used to initialize sentinel-terminated arrays.
```zig
const std = @import("std");
const assert = std.debug.assert;
const expect = std.testing.expect;
test "initialize sentinel-terminated array" {
    // the sentinel does not need to match the value 
    const arr: [2:0]u8 = @splat(10);
    try expect(arr[0] == 10);
    try expect(arr[1] == 10);
    try expect(arr[2] == 0);
}
test "initialize runtime array" {
    var runtime_known: u8 = undefined;
    runtime_known = 123;
    // the operand can be runtime-known, giving a runtime-known array
    const arr: [2]u8 = @splat(runtime_known);
    try expect(arr[0] == 123);
    try expect(arr[1] == 123);
}
test "initialize zero-length sentinel-terminated array" {
    var runtime_known: u8 = undefined;
    runtime_known = 123;
    const arr: [0:10]u8 = @splat(runtime_known);
    // the operand was runtime-known, but since the array length was zero, the result is comptime-known
    comptime assert(arr[0] == 10);
}
```

**Pull request merged**